### ISSUE TITLE
Allow plugins to patch in a config file

### ISSUE THREAD

**Issue by Cooldude2606** [clusterio/clusterio#489]
TL;DR Modules should have access to a config file which has been patched in by a Plugin to avoid needing to use large RCON commands on server startup.

### Details

Most plugins which I want to create use modules which require a configuration file to be present before the game starts. These config files are used to enable/disable certain features of a module or to list the scope the module works over. An example would be a module that creates permission groups in game and requires a list of all the permissions which are allowed, such a file can easily be over 2k character.

### Proposed Implementation

- Modules will have access to a file named `config.lua` at the root of the module. eg `modules/player_auth/config.lua`
- This file is created during save patching using the template `return game.json_to_table("${JSON.stringify(config_object)}")`
- The config object is a property of the module object and is set when by the plugin during `prepareSave` on an instance.
- The plugin sets the property by returning a value from a new hook named `onModuleConfigPatch`

### Limitations

The proposed implementation only covers the creation of a config file during save patching, it does not provide any standard for updating the config while an instance is running. I do not see this as an issue at present as plugins already have the tools required to push config changes through `onInstanceConfigFieldChanged` and `sendRcon` although there could be a method to achieve this later.

If the name `config.lua` is enforced then this would prevent modules from having a file of this name within their root. There are currently no modules which do this so I do not see this as an issue, and because the project is still open for breaking changes this is an acceptable reservation to make.

The return value of the new hook has the abilty to cause errors either in game or in the plugin. A decision should be made on if `game.json_to_table` should be allowed to error, or if validation should be performed before config patching and have a warning emited by the plugin.

### Alternatives

This feature is not implemented and plugins are required to send large RCON commands on instance startup.
Large RCON commands on startup are bad, and if many plugins start doing this then this becomes a bigger issue.

The config file be created in a new dir rather than the root of the module. eg `config/player_auth.json`
This would remove the restriction on modules from having a `config.lua` in their root, but with the added complexity of having a second dir being managed by the save patcher.

Allow plugins to patch arbitrary code into files rather than just a JSON object.
This would be a complex feature to implement when passing a json object to static code covers the vast majority of use cases.

Do not add a new hook to plugins and only use the declared instance config.
The instance config could contain sensitive infomation which should not be included in the distributed save file. Furthermore, some config values might not originate from the declared config.

**Comment by Hornwitser** [comment]
The only part I'm not entirely sold on is calling it for "config" when it's an arbitrary data blob provided by the plugin. We have instance, controller, and host config which is implemented through a common config system. I have thought about extending this config system to include per player configs, and I think it would be very confusing if module config.lua is something completely different.

Do you think you could come up with a different name that isn't confused with the existing usage of config in Clusterio? Like data sync, primordials, save patch data, data store, or something similar? It would be up to the plugin to figure out how to keep it in sync with whatever data source it's using for it.

> Allow plugins to patch arbitrary code into files rather than just a JSON object.
> This would be a complex feature to implement when passing a json object to static code covers the vast majority of use cases.

I don't think this would be that complicated, now that the TypeScript porting is mostly done I think I could implement a simple module system that lets you modify the files patched through an API like this in an afternoon:

```ts
class InstancePlugin /* extends ... */ {
    onModuleLoaded(module: Module) {
        module.files.set(
            "initial_data.lua",
            `return game.json_to_table("${escapeString(JSON.stringify(this.data)}")`
        );
    }
}
```

Would that be more desirable?

> This feature is not implemented and plugins are required to send large RCON commands on instance startup.
> Large RCON commands on startup are bad, and if many plugins start doing this then this becomes a bigger issue.

I don't find this to be true, as long as no player manages to join the server before the initial commands are sent there is no issue with sending MB sized commands at startup as large commands execute instantaneously when no players are online. If a player *do* manage to join in that tiny window of time between startup and the command being sent there will be issues with it taking a long time to be transferred.

**Comment by Danielv123** [comment]
I agree with hornwitser's last point. There is no problem with sending MB sized commands on startup. There is however a problem with designing a complicated solution to adaptively patch in module content on startup

* Complexity
* extra maintenance burden
* Less flexible in use than current solution - can't easily change config at runtime
* errors in the syntax of the loaded code causes server to fail to start

I'd rather see a system to integrate the existing config system better with modules/plugin lua code.
